repos/libports/src/lib/pthread/rwlock.cc

```diff
@@ -0,0 +1,156 @@
+/*
+ * \brief  POSIX readers/writer lock (rwlock) implementation
+ * \author Alexander Senier
+ * \date   2018-01-25
+ *
+ */
+
+/* Genode includes */
+#include <base/log.h>
+#include <base/lock.h>
+#include <base/lock_guard.h>
+
+/* Libc includes */
+#include <pthread.h>
+
+/*
+ * A reader-preferring implementation of a readers-writer lock as described
+ * in Michael Raynal, "Concurrent Programming: Algorithms, Principles, and
+ * Foundations", ISBN 978-3-642-32026-2, page 75
+ */
+
+extern "C" {
+
+	/*
+	 * This class is named 'struct pthread_rwlock' because the 'pthread_rwlock_t'
+	 * type is defined as 'struct rwlock*' in 'sys/_pthreadtypes.h'
+	 */
+	struct pthread_rwlock
+	{
+	private:
+
+		typedef enum { RWLOCK_UNLOCKED, RWLOCK_READ, RWLOCK_WRITE } State;
+
+		/*
+		 * As POSIX defines only a single unlock operation, we need to track
+		 * the state of the lock to either perform an unlock operation for
+		 * a reader lock or a writer lock.
+		 */
+		State _state = RWLOCK_UNLOCKED;
+		Genode::Lock _nbr_mutex {};
+		Genode::Lock _global_mutex {};
+		int _nbr = 0;
+
+		void _begin_read()
+		{
+			Genode::Lock_guard<Genode::Lock> guard(_nbr_mutex);
+			++_nbr;
+			if (_nbr == 1) {
+				_global_mutex.lock();
+			}
+		}
+
+		void _end_read()
+		{
+			Genode::Lock_guard<Genode::Lock> guard(_nbr_mutex);
+			_nbr--;
+			if (_nbr == 0) {
+				_global_mutex.unlock();
+			}
+		}
+
+		void _begin_write() { _global_mutex.lock(); }
+
+		void _end_write() { _global_mutex.unlock(); }
+
+		public:
+			int rdlock()
+			{
+				if (_state != RWLOCK_UNLOCKED) {
+					Genode::warning("Read lock on locked rwlock (state=", int(_state), ")");
+					return -1;
+				}
+
+				_state = RWLOCK_READ;
+				_begin_read();
+				return 0;
+			}
+
+			int wrlock()
+			{
+				if (_state != RWLOCK_UNLOCKED) {
+					Genode::warning("Write lock on locked rwlock (state=", int(_state), ")");
+					return -1;
+				}
+				_state = RWLOCK_WRITE;
+				_begin_write();
```

Comment:
I don't know the article you refer to but couldn't there be preemption before each statement in `rdlock()` and  `wrlock()`? So, how is ensured that `_state` matches the actual lock owner (`READ` or `WRITE`)? Is this property important because it is used on `unlock()`?